src/Microsoft.Extensions.OData.Migration/UriSegmentTranslator.cs

```diff
@@ -0,0 +1,243 @@
+
+            // Use corresponding V4 type to look up corresponding NavigationProperty
+            IEdmNavigationProperty v4navigationProperty = v4Type.FindProperty(segment.NavigationProperty.Name) as IEdmNavigationProperty;
+               
+            // Separately, look up navigation source (entity set)
+            IEdmNavigationSource v4navigationSource = v4model.FindDeclaredNavigationSource(segment.EntitySet.Name);
+
+            return new NavigationPropertySegment(v4navigationProperty, v4navigationSource);
+        }
+
+        /// <summary>
+        /// Translates a V3 EntitySetSegment (e.g. "/Boxes") to a V4 EntitySetSegment
+        /// </summary>
+        /// <param name="segment">OData V3 EntitySetSegment</param>
+        /// <returns>OData V4 EntitySetSegment</returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.EntitySetSegment segment)
+        {
+            IEdmEntitySet v4entitySet = v4model.FindDeclaredEntitySet(segment.EntitySet.Name);
+            EntitySetSegment v4segment = new EntitySetSegment(v4entitySet);
+            return v4segment;
+        }
+
+        /// <summary>
+        /// Translates a V3 KeySegment (e.g. "/Boxes(1)") to a V4 KeySegment
+        /// </summary>
+        /// <param name="segment">OData V3 KeySegment</param>
+        /// <returns>OData V4 KeySegment</returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.KeySegment segment)
+        {
+            IEdmNavigationSource v4navigationSource = v4model.FindDeclaredNavigationSource(segment.EntitySet.Name);
+            IEdmEntityType v4type = (IEdmEntityType)v4model.FindType(segment.EdmType.FullTypeName());
+            KeySegment v4segment = new KeySegment(segment.Keys, v4type, v4navigationSource);
+            return v4segment;
+        }
+
+        /// <summary>
+        /// Translates at V3 PropertySegment (e.g. "/Boxes(1)/Name") into a V4 PropertySegment
+        /// </summary>
+        /// <param name="segment">OData V3 PropertySegment</param>
+        /// <returns>OData V4 PropertySegment</returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.PropertySegment segment)
+        {
+            // Extract the DeclaringType (e.g. EntitySet) name from V3 PropertySegment
+            string v3typeName = segment.Property.DeclaringType.FullTypeName();
+
+            // Look up V4 Structured Type (guaranteed if found) in V4 model
+            IEdmStructuredType v4Type = v4model.FindType(v3typeName) as IEdmStructuredType;
+
+            // Look up corresponding Property in V4 Type by name
+            IEdmStructuralProperty v4Property = v4Type.FindProperty(segment.Property.Name) as IEdmStructuralProperty;
+
+            return new PropertySegment(v4Property);
+        }
+
+        /// <summary>
+        /// Translates a V3 OperationSegment to V4 OperationSegment by locating all matching function imports in the v4 model

```

Comment:
Example(s) would be very helpful in explaining the logic of the function

Reply:
I will look for examples and if I'm unable to find the by next meeting with Sam I'll ask him!